tools/codegen/README.md

```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
+- `--controller-gen` - optionally use a particular controller-gen binary. When not specified, the tool will use the
+  built in generator.
+  Note, you must use a `controller-gen` built from the [OpenShift fork](https://github.com/openshift/kubernetes-sigs-controller-tools) of controller-tools.
+- `--required-feature-sets` - optionally generate based on the OpenShift feature sets annotations.
+  This will update only CRDs with a matching value for the `release.openshift.io/feature-set` annotation.
+
+For example, to generate only TechPreviewNoUpgrade versions of CRDs:
+```
+codegen schemapatch --base-dir /go/src/github.com/openshift/api --api-group-versions apps.openshift.io/v1,config.openshift.io/v1,operator.openshift.io/v1 --require-feature-sets TechPreviewNoUpgrade
+```
+
+[controller-tools]: https://github.com/kubernetes-sigs/controller-tools
+[yaml-patch]: https://github.com/vmware-archive/yaml-patch
+
+### Swaggerdocs
+
+To generate SwaggerDoc functions for types within an API, use the `swaggerdocs` subcommand.
+
+This generator inspects the documentation within the API defintions and generates functions
+that return the Go documenation for each struct and field within an API.
+
+The generator can also be used to verify that each API field and struct has appropriate
+documentation, using the `--swagger:enforce-comments` and `--verify` flags in conjunction.
+
+These additional arguments may be provided when using the `swaggerdocs` generator:
+- `--swagger:output-file-name` -  Defines the file name to use for the swagger generated docs for each group version.
+  When omitted, defaults to `zz_generated.swagger_doc_generated.go`.
+- `--swagger:enforce-comments` - Defines whether the generator should enforce that all fields have a comment or not.
+  Only effective when combined with `--verify`. When not specified, a warning will be emitted for each missing comment.
+
+## Using configuration files
+
+The `codegen` utility will search for a file called `.codegen.yaml` at the API group level.
+These files allow individual API groups to configure the generation of their paritcular API files.
+
+It also enables the enablement and disablement of generators with a goal that the usage of the generator
+should require only the `--base-dir` argument.
+
+The schema of the configuration file is outlined below, all fields are optional and default values will apply
+when they are omitted.
+
+Each generator option relates to a flag attached to the generator, review the [generators](#generators) section
+for more details on each option within the config file structure.
+
+```yaml
+# Configuration for the compatibility generator.
+compatibility:
+  enabled: true # Defaults to true, set to false to disable.
+# Configuration for the schemapatch generator.
+schemapatch:
+  enabled: true # Defaults to true, set to false to disable.
+  requiredFeatureSets: # Each entry will be matched against the value of the required feature set annotation.
```

Comment:
What is the default?  I'd like to avoid having any package in openshift need to set this.

Reply:
The default is that it only generates manifests that don't have an annotation at all, which is the same behaviour as `controller-gen`. If we want to avoid users having to configure this, then we will have to add discovery of the required feature sets

We should collapse this conversation into https://github.com/openshift/api/pull/1343/files#r1025296179